src/core/diamond.cpp

```diff
@@ -0,0 +1,156 @@
+/*
+  Copyright (C) 2010-2018 The ESPResSo project
+  Copyright (C) 2002,2003,2004,2005,2006,2007,2008,2009,2010
+    Max-Planck-Institute for Polymer Research, Theory Group
+
+  This file is part of ESPResSo.
+
+  ESPResSo is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  ESPResSo is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/** \file
+    This file contains everything needed to create a start-up configuration
+    of (partially charged) diamond structure polymer chains with counterions
+    and salt molecules, assigning velocities to the particles and
+    cross-linking the polymers if necessary.
+
+    The corresponding header file is diamond.hpp.
+*/
+
+#include <cmath>
+#include <cstddef>
+#include <cstdio>
+#include <cstdlib>
+#include <cstring>
+
+#include "PartCfg.hpp"
+#include "bonded_interactions/bonded_interaction_data.hpp"
+#include "communication.hpp"
+#include "constraints.hpp"
+#include "constraints/ShapeBasedConstraint.hpp"
+#include "debug.hpp"
+#include "diamond.hpp"
+#include "global.hpp"
+#include "integrate.hpp"
+#include "polymer.hpp"
+#include "random.hpp"
+
+/*************************************************************
+ * Functions                                                 *
+ * ---------                                                 *
+ *************************************************************/
+
+int create_counterions(PartCfg &partCfg, int const N_CI, int part_id,
+                       int const mode, double const shield, int const max_try,
+                       double const val_CI, int const type_CI) {
+  Utils::Vector3d pos;
+
+  int cnt1 = 0;
+  int max_cnt = 0;
+  for (int n = 0; n < N_CI; ++n) {
+    for (cnt1 = 0; cnt1 < max_try; ++cnt1) {
+      pos[0] = box_l[0] * d_random();
+      pos[1] = box_l[1] * d_random();
+      pos[2] = box_l[2] * d_random();
+      if ((mode != 0) or (mindist(partCfg, pos) > shield))
+        break;
+    }
+    if (cnt1 >= max_try)
+      throw std::runtime_error("Too many failed attpts finding valid position.");
+    if (place_particle(part_id, &pos[0]) == ES_PART_ERROR)
+      throw std::runtime_error("Failed to place particle.");
+    set_particle_q(part_id, val_CI);
+    set_particle_type(part_id, type_CI);
+
+    part_id++;
+    max_cnt = std::max(cnt1, max_cnt);
+  }
+  POLY_TRACE(printf(" %d->%d \n", cnt1, max_cnt));
+  if (cnt1 >= max_try)
+    throw std::runtime_error("Too many failed attpts finding valid position.");
```

Comment:
this branch will never be visited because the same condition at lines 68-69 throws